File: BE/api/inquiry.py
"""
문의 처리 라우터.
"""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from BE.core.schemas import (
    InquiryRequest, InquirySubmitResponse, InquiryStatusResponse, InquiryStatus,
    ReviewRequest, PendingInquiryResponse, RuleUpdateRequest, DeptChangeRequest,
)
from BE.api.pipeline import process_inquiry
from BE.db.database import get_db
from BE.db.models import InquiryRecord
from BE.db import crud
from BE.core.deps import get_current_user, require_staff, require_master

router = APIRouter(prefix="/api", tags=["inquiry"])


def _to_status_response(record: InquiryRecord) -> InquiryStatusResponse:
    if record.reviewed:
        return InquiryStatusResponse(
            id=record.id,
            created_at=record.created_at.isoformat(),
            original_text=record.original_text,
            department=record.department,
            priority=record.priority,
            status=InquiryStatus.ANSWERED,
            answer=record.final_answer or record.answer_draft,
        )
    return InquiryStatusResponse(
        id=record.id,
        created_at=record.created_at.isoformat(),
        original_text=record.original_text,
        department=record.department,
        priority=record.priority,
        status=InquiryStatus.PENDING,
        answer=None,
    )


@router.post("/inquiry", response_model=InquirySubmitResponse, status_code=status.HTTP_201_CREATED)
def create_inquiry(
    req: InquiryRequest,
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    response = process_inquiry(req.text)
    try:
        record = crud.save_inquiry(db, response, user_id=int(user["sub"]))
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="문의 등록 중 오류가 발생했습니다. 잠시 후 다시 시도해 주세요.",
        )
    return InquirySubmitResponse(id=record.id, created_at=record.created_at.isoformat())


@router.get("/inquiry/my", response_model=list[InquiryStatusResponse])
def list_my_inquiries(db: Session = Depends(get_db), user: dict = Depends(get_current_user)):
    records = crud.get_user_inquiries(db, user_id=int(user["sub"]))
    return [_to_status_response(r) for r in records]


@router.get("/inquiry/my/{inquiry_id}", response_model=InquiryStatusResponse)
def get_my_inquiry(
    inquiry_id: int, db: Session = Depends(get_db), user: dict = Depends(get_current_user)
):
    record = crud.get_user_inquiry_by_id(db, inquiry_id, user_id=int(user["sub"]))
    if record is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="문의를 찾을 수 없습니다.")
    return _to_status_response(record)


def _to_pending_response(r: InquiryRecord) -> PendingInquiryResponse:
    return PendingInquiryResponse(
        id=r.id,
        created_at=r.created_at.isoformat(),
        original_text=r.original_text,
        inquiry_type=r.inquiry_type,
        key_request=r.key_request,
        confidence=r.confidence,
        domain=r.domain,
        category=r.category,
        department=r.department,
        department_certain=r.department_certain,
        department_note=r.department_note,
        urgent_reason=r.urgent_reason,
        priority=r.priority,
        answer_draft=r.answer_draft,
        answer_confidence=r.answer_confidence,
        retrieved=r.retrieved_docs or [],
        used_llm=r.used_llm,
        llm_error=r.llm_error,
        dept_change_requested=r.dept_change_requested,
        dept_change_reason=r.dept_change_reason,
        dept_change_suggested=r.dept_change_suggested,
        dept_change_requested_by=r.dept_change_requested_by,
    )


@router.get("/inquiry/pending", response_model=list[PendingInquiryResponse])
def list_pending_inquiries(db: Session = Depends(get_db), user: dict = Depends(require_staff)):
    """
    master는 전체 부서 큐를 본다(재배정 판단을 위해 필요). 일반 staff는 자기
    부서로 배정된 것만 본다. staff인데 department가 아직 지정 안 됐으면
    (운영자가 SQL로 role만 올리고 department는 안 지정한 경우) 아무 부서
    문의도 잘못 보여주지 않도록 안전하게 빈 목록을 반환한다.
    """
    if user.get("role") == "master":
        records = crud.get_pending_inquiries(db)
    else:
        dept = user.get("department")
        if not dept:
            return []
        records = crud.get_pending_inquiries(db, department=dept)
    return [_to_pending_response(r) for r in records]


@router.patch("/inquiry/{inquiry_id}/review", response_model=InquiryStatusResponse)
def review_inquiry(
    inquiry_id: int,
    req: ReviewRequest,
    db: Session = Depends(get_db),
    user: dict = Depends(require_staff),
):
    record = crud.mark_reviewed(db, inquiry_id, reviewed_by=user["email"], final_answer=req.final_answer)
    if record is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="문의를 찾을 수 없습니다.")
    return _to_status_response(record)


@router.patch("/inquiry/{inquiry_id}/request-department-change", response_model=PendingInquiryResponse)
def request_department_change(
    inquiry_id: int,
    req: DeptChangeRequest,
    db: Session = Depends(get_db),
    user: dict = Depends(require_staff),
):
    """
    staff가 "이 문의는 우리 부서 업무가 아닌 것 같다"고 재배정을 요청.
    실제 재배정 권한은 master에게만 있으므로(PATCH /rule 참고), 이 엔드포인트는
    부서를 직접 바꾸지 않고 검토대기 큐(GET /pending)에 요청 표시만 남긴다.
    master가 큐에서 이 표시를 보고 PATCH /rule로 실제 재배정하면, 그 순간
    이 요청 표시는 자동으로 해소(초기화)된다(crud.update_rule 참고).
    """
    record = crud.request_department_change(
        db, inquiry_id, reason=req.reason, requested_by=user["email"],
        suggested_department=req.suggested_department,
    )
    if record is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="문의를 찾을 수 없거나, 이미 검토완료된 문의라 요청할 수 없습니다.",
        )
    return _to_pending_response(record)


@router.patch("/inquiry/{inquiry_id}/rule", response_model=PendingInquiryResponse)
def update_inquiry_rule(
    inquiry_id: int,
    req: RuleUpdateRequest,
    db: Session = Depends(get_db),
    user: dict = Depends(require_master),
):
    record = crud.update_rule(db, inquiry_id, priority=req.priority, department=req.department)
    if record is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="문의를 찾을 수 없습니다.")
    return _to_pending_response(record)

File: FE/src/types/inquiry.ts
// 백엔드 응답 스펙 (v4 기준 — 검토 화면 데이터 확장 + 권한 3단계)
//
// 흐름:
//   사용자: POST /api/inquiry (등록)          -> 접수 확인만 받음 (답변 내용 모름)
//   사용자: GET  /api/inquiry/my (내 문의함)   -> 로그인 기반 자동 필터링, 검토 끝난 건만 답변 보임
//   담당자(staff+): GET /api/inquiry/pending   -> 분류·근거문서·AI 초안·확신도 전부 확인
//   담당자(staff+): PATCH /api/inquiry/{id}/review -> 그대로 승인 또는 수정 후 승인
//   최고관리자(master만): PATCH /api/inquiry/{id}/rule -> 부서/우선순위 재배정
//
// 권한: general < staff < master (계층형. master는 staff가 하는 것도 다 가능)

export type InquiryType =
  | "경력인증"
  | "신청"
  | "변경"
  | "취소·환불"
  | "오류·장애"
  | "일반문의"
  | "긴급문의"
  | "미분류";

export type Priority = "긴급" | "높음" | "보통";

/** 사용자 쪽에서 보이는 문의 처리 상태 (담당자 검토 전/후) */
export type RecordStatus = "pending" | "answered";

/** 답변이 근거로 얼마나 뒷받침되는지 (담당자 검토용) */
export type AnswerConfidence = "sufficient" | "partial" | "insufficient";

/** "admin"(행정) 또는 "technical"(기술질의) */
export type InquiryDomain = "admin" | "technical";

/** RAG 검색으로 찾은 근거 문서 한 건 */
export interface RetrievedDoc {
  content: string;
  /** "파일명 > 질문" 형태의 출처 */
  source: string;
  /** 유사도 점수. 0.4 이하면 "근거 약함" */
  score: number;
  /** AI가 이 문서를 실제로 답변 생성에 사용했는지. 기본값 true */
  selected: boolean;
}

/** POST /api/inquiry 의 요청 바디 */
export interface InquiryRequest {
  text: string;
}

/** POST /api/inquiry 의 응답 바디. 이제 접수 확인만 온다 (답변은 안 옴). */
export interface InquiryCreateResponse {
  id: number;
  message: string;
  created_at: string;
}

/**
 * GET /api/inquiry/my, GET /api/inquiry/my/{id} 의 응답.
 * status가 "pending"이면 answer는 항상 null.
 */
export interface MyInquiry {
  id: number;
  created_at: string;
  original_text: string;
  department: string;
  priority: Priority;
  status: RecordStatus;
  answer: string | null;
}

/**
 * GET /api/inquiry/pending 의 응답 (staff 이상 전용).
 * v4에서 분류·근거문서·LLM 실패 여부가 전부 복원되었다.
 */
export interface PendingInquiry {
  id: number;
  created_at: string;
  original_text: string;
  inquiry_type: InquiryType;
  /** 핵심 요청 한 줄 요약 */
  key_request: string;
  /** 분류 확신도 0~1. 0.5 미만이면 "분류 불확실" 표시 */
  confidence: number;
  domain: InquiryDomain;
  department: string;
  priority: Priority;
  answer_draft: string;
  answer_confidence: AnswerConfidence;
  retrieved: RetrievedDoc[];
  used_llm: boolean;
  /** null 이면 정상, 값이 있으면 LLM 생성 실패 */
  llm_error: string | null;

  // v4 델타 — staff의 부서변경 요청 표시. master가 /rule 로 재배정하면
  // 서버가 이 4개를 자동으로 초기화한다 (프론트에서 따로 지울 필요 없음).
  /** staff가 부서변경을 요청했는지 */
  dept_change_requested: boolean;
  dept_change_reason: string | null;
  dept_change_suggested: string | null;
  /** 요청한 staff의 이메일 */
  dept_change_requested_by: string | null;
}

/** PATCH /api/inquiry/{id}/review 의 요청 바디 */
export interface ReviewRequest {
  /** null(또는 생략)이면 AI 초안 그대로 승인, 값이 있으면 그 내용으로 교체해서 승인 */
  final_answer: string | null;
}

/** PATCH /api/inquiry/{id}/review 의 응답. GET /api/inquiry/my 항목과 같은 형식. */
export type ReviewResponse = MyInquiry;

/**
 * PATCH /api/inquiry/{id}/rule 의 요청 바디 (master 전용).
 * 둘 다 선택 필드 — 하나만 보내면 그것만 바뀐다.
 */
export interface RuleUpdateRequest {
  priority?: Priority;
  department?: string;
}

/**
 * PATCH /api/inquiry/{id}/rule 의 응답.
 * ⚠️ v4 델타에서 바뀜: 이전엔 MyInquiry(간단 정보)였는데, 이제는
 * PendingInquiry(분류·근거문서·confidence 등 전체 정보)로 온다.
 */
export type RuleUpdateResponse = PendingInquiry;

/** PATCH /api/inquiry/{id}/request-department-change 의 요청 바디 (staff 이상) */
export interface DeptChangeRequest {
  reason: string;
  /** 선택. 없으면 null */
  suggested_department?: string | null;
}

/** PATCH /api/inquiry/{id}/request-department-change 의 응답. /pending 항목과 동일. */
export type DeptChangeResponse = PendingInquiry;